platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -76,44 +86,83 @@ private void setConvention(CollectionSupplier<T, C> unguardedConvention) {
      */
     protected abstract C emptyCollection();
 
+    private CollectionPropertyConfigurer<T> getConventionValue() {
+        return new ConventionConfigurer();
+    }
+
+    public CollectionPropertyConfigurer<T> getExplicitValue() {
+        return new ExplicitValueConfigurer();
+    }
+
+    @Override
+    public HasMultipleValues<T> withActualValue(Action<CollectionPropertyConfigurer<T>> action) {
+        action.execute(getActualValue());
+        return this;
+    }
+
+    @Override
+    public HasMultipleValues<T> withActualValue(Closure<Void> action) {
+        ConfigureUtil.configure(action, getActualValue());
+        return this;
+    }
+
+    private CollectionPropertyConfigurer<T> getActualValue() {
+        if (isExplicit() || isDefaultConvention()) {
+            return getExplicitValue();
+        }
+        return getConventionValue();
+    }
```

Comment:
:question: The fact that `withActualValue` may update the convention instead of the value should be explicitly documented, I think.

Reply:
🤔 I missed that `withActualValue` would actually update the convention value (instead of using it just a shallow copy as the starting point for incremental updates)

Reply:
I am changing that behavior so `withActualValue`  commits the convention value to the explicit value and always runs against the explicit value.